### PR TITLE
[4/X][Pipeline] JSON serialization for serve dag nodes

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

Added JSON serde for all DAGNode types needed with tests on ray core dag as well as serve dag. See code inline comments for behavior and assumption for each.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/pipeline/json_serde.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            elif isinstance(dag_node, FunctionNode):
+                body = dag_node._body
+                import_path = f"{body.__module__}.{body.__qualname__}"
+            elif isinstance(dag_node, (DeploymentNode, DeploymentMethodNode)):
+                result_dict.update(
+                    {"deployment_name": dag_node._deployment.name}
+                )
+                if isinstance(dag_node._deployment._func_or_class, str):
+                    # We're processing a deserilized JSON node where import_path
+                    # is dag_node body.
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:67]: please make the `import_path` logic a standalone utility and add unit tests for it.

File: python/ray/serve/pipeline/json_serde.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+            return result_dict
+        else:
+            # Let the base class default method raise the TypeError
+            return json.JSONEncoder.default(self, obj)
+
+
+def dagnode_from_json(input_json: Any) -> Union[DAGNode, RayServeHandle, Any]:
+    """JSON serialization is only used and enforced in ray serve from ray core
+    API authored DAGNode(s).
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:91]: update this to describe what it does and what the assumptions are

File: python/ray/serve/pipeline/json_serde.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            return InputNode()
+
+        # Pre-order JSON deserialization
+        args = json.loads(input_json["args"], object_hook=dagnode_from_json)
+        kwargs = json.loads(input_json["kwargs"], object_hook=dagnode_from_json)
+        # .options() should not contain any DAGNode type
+        options = json.loads(input_json["options"])
+        other_args_to_resolve = json.loads(
+            input_json["other_args_to_resolve"], object_hook=dagnode_from_json
+        )
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:118]: don't do all of this handling in the `except` block, it'll lead to nasty stack traces

File: python/ray/serve/utils.py [hunk 4]
```diff
@@ -102,9 +107,69 @@ def default(self, o):  # pylint: disable=E0202
...
 
 
+class ServeHandleEncoder(json.JSONEncoder):
+    """JSON encoder for RayServeHandle and RayServeSyncHandle. Use to enforce
+    JSON serialization of deployment init args & kwargs to faciliate serve
+    pipeline deployment as well as operationaling serve.
+    """
+
+    def default(self, obj):
+        # Import RayServeHandle in utils file lead to import errors
```
Comment (edoakes) [inline python/ray/serve/utils.py:117]: unit tests for these please

Comment (jiaodong) [reply python/ray/serve/utils.py:117]: I think the simplest thing is to extend our existing `test_handles.py` to do a full round of JSON serde and ensure the outcome is still the same. I've updated two existing tests, lmk if you have other good ones in mind for better coverage

Comment (edoakes) [reply python/ray/serve/utils.py:118]: yep that sounds good

File: python/ray/serve/constants.py [hunk 5]
```diff
@@ -74,3 +74,8 @@
 #: Number of times in a row that a replica must fail the health check before
 #: being marked unhealthy.
 REPLICA_HEALTH_CHECK_UNHEALTHY_THRESHOLD = 3
+
+# Key used to idenfity given json represents a serialized RayServeHandle
+SERVE_ASYNC_HANDLE_JSON_KEY = "__SerializedServeHandle__"
+# Key used to idenfity given json represents a serialized RayServeSyncHandle
+SERVE_SYNC_HANDLE_JSON_KEY = "__SerializedServeSyncHandle__"
```
Comment (edoakes) [inline python/ray/serve/constants.py:81]: If the JSON is a dictionary anyways, probably preferable to have a single key and have a `sync` flag that indicates if it's sync or async

File: python/ray/serve/pipeline/deployment_node.py [hunk 6]
```diff
@@ -13,7 +13,9 @@ class DeploymentNode(DAGNode):
 
     def __init__(
         self,
-        func_or_class,
+        # For serve structured deployment, deployment body can be import path
+        # to the class or function instead.
+        func_or_class: Union[Callable, str],
```
Comment (edoakes) [inline python/ray/serve/pipeline/deployment_node.py:18]: :'(

Comment (edoakes) [reply python/ray/serve/pipeline/deployment_node.py:18]: will be unified to only an import path soon

File: python/ray/serve/pipeline/json_serde.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        - All modules such as class or functions need to be visible and
+            importable on top of its file, and can be resolved via a fully
+            qualified import_path.
+        - No DAGNode instance should appear in bound .options(), which should be
+            JSON serializable with default encoder.
+    """
+
+    def default(self, obj):
+        # TODO: (jiaodong) Have better abtraction for users to extend and use
+        # JSON serialization of their own object types
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:47]: not sure this is desirable, can be done in application layer

File: python/ray/serve/pipeline/json_serde.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        # For replaced deployment handles used as init args or kwargs.
+        if isinstance(obj, RayServeHandle):
+            return json.dumps(obj, cls=ServeHandleEncoder)
+        # For all other DAGNode types.
+        if isinstance(obj, DAGNode):
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:52]: prefer to use `if-elif-else` chain and only have a single return at the bottom of the function to simplify the control flow

File: python/ray/serve/pipeline/json_serde.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+                # Placeholder for better ordering
+                "import_path": "",
+                "args": json.dumps(args, cls=DAGNodeEncoder),
+                "kwargs": json.dumps(kwargs, cls=DAGNodeEncoder),
+                # .options() should not contain any DAGNode type
+                "options": json.dumps(options),
+                "other_args_to_resolve": json.dumps(
+                    other_args_to_resolve, cls=DAGNodeEncoder
+                ),
+            }
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:71]: hmm can you put this in a separate utility? e.g., `dag_node.to_json()`? You can then add a corresponding. `DAGNode.from_json()` classmethod

File: python/ray/serve/pipeline/json_serde.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+            elif isinstance(dag_node, ClassMethodNode):
+                parent_class_node = dag_node._parent_class_node
+                body = dag_node._parent_class_node._body.__ray_actor_class__
+                import_path = f"{body.__module__}.{body.__qualname__}"
+                result_dict.update(
+                    {
+                        "method_name": dag_node._method_name,
+                        "parent_class_node_stable_uuid": parent_class_node._stable_uuid,
+                    }
+                )
```
Comment (edoakes) [inline python/ray/serve/pipeline/json_serde.py:90]: the to_json / from_json I suggested would also make this cleaner -- each node type has its own impl

File: python/ray/serve/pipeline/json_serde.py (hunk 1 above)
Comment (jiaodong) [reply python/ray/serve/pipeline/json_serde.py]: i moved these to each DAGNode type now, will address in later diff before beta when we polish generated yaml part. We can also cover with some remote runtime_env tests to properly test all import_path related items.

File: python/ray/serve/pipeline/deployment_method_node.py [hunk 11]
```diff
@@ -94,3 +95,40 @@ def __str__(self) -> str:
...
+        json_dict[DAGNODE_TYPE_KEY] = DeploymentMethodNode.__name__
+        json_dict["deployment_name"] = self._deployment.name
+        json_dict["method_name"] = self._method_name
+
+        if isinstance(self._deployment._func_or_class, str):
+            # We're processing a deserilized JSON node where import_path
+            # is dag_node body.
+            json_dict["import_path"] = self._deployment._func_or_class
+        else:
+            body = self._deployment._func_or_class.__ray_actor_class__
```
Comment (edoakes) [inline python/ray/serve/pipeline/deployment_method_node.py:110]: don't access private attrs

**Review by edoakes — COMMENTED** [review]
A few more stylistic comments.

Can you also add a test for the case where we fail to serialize to JSON? This should raise an exception with a helpful error message.

**Comment by jiaodong** [comment]
looks like `test_json_serde` failed on windows only but passed on python ones .. cc: @simon-mo any tips about reproducing it ?

**Review by edoakes — APPROVED** [review]
Nice, much better now :)

**Pull request merged**